keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1037 @@
+unnecessary manner.
+
+Lacking the knob to flag Pods as scheduling-ready/unready wastes scheduling cycles on retrying
+Pods that are determined to be unschedulable. As a result, it delays the scheduling of other Pods,
+and also lowers the overall scheduling throughput. Moreover, it imposes restrictions to vendors to
+develop some in-house features (such as hierarchical quota) natively.
+
+This proposal describes APIs and mechanics to allow users/controllers to control when a pod is
+ready to be considered for scheduling.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Define an API to mark Pods as scheduling-paused/unpaused.
+- Design a new Enqueue extension point to customize Pod's queueing behavior.
+- A default enqueue plugin to honor the new API semantics.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Update the Pod's condition if it carries `.spec.schedulingPaused=true`.
+- Focus on in-house use-cases of the Enqueue extension point.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose a new field `.spec.schedulingPaused` to the Pod API. The field is defaulted to `false`.
+
+For Pods carrying `.spec.schedulingPaused=true`, they will be "parked" in scheduler's internal
+unschedulablePods pool, and only get tried when the field is set to `false`.
+
+Practically, this field can be initialized to `true` by a MutatingWebhook, and be flipped to `false`
+by external integrators when certain criteria is met.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an orchestrator developer, such as a dynamic quota manager, I have the full picutre to know when
+Pods are scheduling-ready; therefore, I want an API to signal to kube-scheduler when to consider a
+Pod for scheduling. The pattern for this story would be to use a mutating webhook to force creating
+pods in a "not-ready to schedule" state that the custom orchestrator changes to ready at a later
+time based on its own evaluations.
+
+#### Story 2
+
+This story is an extension of the previous one in that more than one custom orchestrator could be
+deployed on the same cluster, therefore they want an API that enables establishing an agreement om
+when a pod is considered ready for scheduling.
+
+#### Story 3
```

Comment:
I suggest we add one more user story, relevant to @pwschuurman's suggestion (it could be 1a, but i want to make it clear the primary goal is scheduling fitness with the side effect of improving others):

```
#### Story 4

A custom workload orchestrator may wish to modify the pod prior to consideration for scheduling, without having to fork or alter the workload controller.  The orchestrator may wish to make time-varying (post-creation) decisions on pod scheduling, perhaps to preserve scheduling constraints, avoid disruption, or prevent co-existence.  

```

Reply:
Thanks! Will udpate.

Reply:
updated.